File: video_compress.py
# video_compress.py
import subprocess
import argparse
import os
import sys
import shutil
from pathlib import Path
from concurrent.futures import ThreadPoolExecutor, as_completed

SUPPORTED_INPUT = ('.mp4', '.mkv', '.avi', '.mov', '.wmv', '.flv', '.webm', '.m4v', '.3gp', '.ogv')
VIDEO_CODECS = {
    'h264': 'libx264',
    'hevc': 'libx265',
    'vp9': 'libvpx-vp9',
    'av1': 'libaom-av1'
}
AUDIO_CODECS = ['aac', 'mp3', 'opus', 'ac3']
PRESETS = ['ultrafast', 'superfast', 'veryfast', 'faster', 'fast', 'medium', 'slow', 'slower', 'veryslow']

def check_ffmpeg():
    if shutil.which("ffmpeg") is None:
        print("Ошибка: ffmpeg не найден. Установите FFmpeg и добавьте в PATH.", file=sys.stderr)
        sys.exit(1)

def build_ffmpeg_cmd(input_file, output_file, video_codec, audio_codec, crf, video_bitrate, audio_bitrate,
                     size, fps, preset, two_pass, normalize, subtitles):
    cmd = ["ffmpeg", "-i", input_file, "-y"]
    
    # Видеокодек
    if video_codec in VIDEO_CODECS:
        cmd.extend(["-c:v", VIDEO_CODECS[video_codec]])
    else:
        cmd.extend(["-c:v", "libx264"])
    
    # Аудиокодек
    if audio_codec in AUDIO_CODECS:
        cmd.extend(["-c:a", audio_codec])
    else:
        cmd.extend(["-c:a", "aac"])
    
    # CRF или битрейт
    if crf is not None:
        cmd.extend(["-crf", str(crf)])
    elif video_bitrate:
        cmd.extend(["-b:v", video_bitrate])
    
    # Битрейт аудио
    if audio_bitrate:
        cmd.extend(["-b:a", audio_bitrate])
    
    # Разрешение
    if size:
        cmd.extend(["-vf", f"scale={size}:flags=lanczos"])
    
    # Частота кадров
    if fps:
        cmd.extend(["-r", str(fps)])
    
    # Пресет
    if preset and preset in PRESETS:
        cmd.extend(["-preset", preset])
    
    # Нормализация громкости
    if normalize:
        cmd.extend(["-af", "loudnorm=I=-16:LRA=11:TP=-1.5"])
    
    # Субтитры
    if subtitles:
        cmd.extend(["-vf", f"subtitles={subtitles}"])
    
    cmd.append(output_file)
    return cmd

def convert_file(input_path, output_path, video_codec, audio_codec, crf, video_bitrate, audio_bitrate,
                 size, fps, preset, two_pass, normalize, subtitles):
    # Двухпроходное кодирование
    if two_pass and video_bitrate:
        # Первый проход
        pass1_cmd = build_ffmpeg_cmd(str(input_path), "/dev/null", video_codec, audio_codec, crf, video_bitrate,
                                     audio_bitrate, size, fps, preset, False, normalize, subtitles)
        pass1_cmd.extend(["-pass", "1", "-f", "null"])
        print(f"  Проход 1/2: {' '.join(pass1_cmd)}")
        subprocess.run(pass1_cmd, check=True, capture_output=True)
        # Второй проход
        pass2_cmd = build_ffmpeg_cmd(str(input_path), str(output_path), video_codec, audio_codec, crf, video_bitrate,
                                     audio_bitrate, size, fps, preset, False, normalize, subtitles)
        pass2_cmd.extend(["-pass", "2"])
        print(f"  Проход 2/2: {' '.join(pass2_cmd)}")
        try:
            subprocess.run(pass2_cmd, check=True, capture_output=True)
            return True
        except subprocess.CalledProcessError as e:
            print(f"  Ошибка: {e.stderr.decode()}", file=sys.stderr)
            return False
    else:
        cmd = build_ffmpeg_cmd(str(input_path), str(output_path), video_codec, audio_codec, crf, video_bitrate,
                               audio_bitrate, size, fps, preset, False, normalize, subtitles)
        try:
            subprocess.run(cmd, check=True, capture_output=True)
            return True
        except subprocess.CalledProcessError as e:
            print(f"  Ошибка: {e.stderr.decode()}", file=sys.stderr)
            return False

def find_video_files(root, recursive):
    files = []
    root_path = Path(root)
    if root_path.is_file() and root_path.suffix.lower() in SUPPORTED_INPUT:
        return [root_path]
    if not root_path.is_dir():
        return []
    pattern = "**/*" if recursive else "*"
    for p in root_path.glob(pattern):
        if p.is_file() and p.suffix.lower() in SUPPORTED_INPUT:
            files.append(p)
    return files

def main():
    parser = argparse.ArgumentParser(description='Конвертер видео (сжатие)')
    parser.add_argument('source', help='Входной файл или папка')
    parser.add_argument('--output', '-o', help='Выходной файл или папка')
    parser.add_argument('--codec', choices=VIDEO_CODECS.keys(), default='h264', help='Видеокодек')
    parser.add_argument('--audio-codec', choices=AUDIO_CODECS, default='aac', help='Аудиокодек')
    parser.add_argument('--crf', type=int, help='CRF (0-51, меньше = лучше качество)')
    parser.add_argument('--video-bitrate', help='Битрейт видео (например, 1000k)')
    parser.add_argument('--audio-bitrate', help='Битрейт аудио (например, 128k)')
    parser.add_argument('--size', help='Разрешение (ШxВ)')
    parser.add_argument('--fps', type=int, help='Частота кадров')
    parser.add_argument('--preset', choices=PRESETS, default='medium', help='Пресет скорости')
    parser.add_argument('--two-pass', action='store_true', help='Двухпроходное кодирование')
    parser.add_argument('--normalize', action='store_true', help='Нормализация громкости')
    parser.add_argument('--subtitles', help='Файл субтитров для встраивания')
    parser.add_argument('--recursive', action='store_true', help='Рекурсивный обход папок')
    parser.add_argument('--threads', type=int, default=4, help='Количество потоков')
    parser.add_argument('--overwrite', action='store_true', help='Перезаписывать файлы')
    parser.add_argument('--formats', action='store_true', help='Показать поддерживаемые форматы')
    args = parser.parse_args()

    if args.formats:
        print("Входные форматы:", ', '.join(SUPPORTED_INPUT))
        print("Видеокодеки:", ', '.join(VIDEO_CODECS.keys()))
        print("Аудиокодеки:", ', '.join(AUDIO_CODECS))
        print("Пресеты:", ', '.join(PRESETS))
        return

    check_ffmpeg()

    if args.crf is not None and args.video_bitrate:
        print("Ошибка: укажите либо --crf, либо --video-bitrate, но не оба.", file=sys.stderr)
        sys.exit(1)

    files = find_video_files(args.source, args.recursive)
    if not files:
        print(f"Не найдено видеофайлов в {args.source}")
        sys.exit(1)

    output_dir = Path(args.output) if args.output else Path("./compressed")
    output_dir.mkdir(parents=True, exist_ok=True)
    total = len(files)
    print(f"Найдено {total} видеофайлов.")
    success = 0

    def worker(input_file):
        nonlocal success
        if args.output and Path(args.output).suffix:
            if len(files) > 1:
                return (False, "Указан один выходной файл, но найдено несколько входных.")
            out_file = Path(args.output)
        else:
            rel = input_file.relative_to(args.source) if Path(args.source).is_dir() else input_file.name
            out_file = output_dir / rel.with_suffix('.mp4')
        if out_file.exists() and not args.overwrite:
            return (False, f"{out_file} уже существует, пропуск.")
        out_file.parent.mkdir(parents=True, exist_ok=True)
        if convert_file(input_file, out_file, args.codec, args.audio_codec, args.crf, args.video_bitrate,
                        args.audio_bitrate, args.size, args.fps, args.preset, args.two_pass,
                        args.normalize, args.subtitles):
            return (True, f"Конвертация {input_file} -> {out_file}")
        return (False, f"Ошибка при конвертации {input_file}")

    with ThreadPoolExecutor(max_workers=args.threads) as executor:
        futures = {executor.submit(worker, f): f for f in files}
        for i, future in enumerate(as_completed(futures), 1):
            ok, msg = future.result()
            print(f"[{i}/{total}] {msg}")
            if ok:
                success += 1

    print(f"Готово! Успешно: {success}, Всего: {total}")

if __name__ == '__main__':
    main()

File: video_compress.rs
// video_compress.rs
use std::fs;
use std::path::{Path, PathBuf};
use std::process::Command;
use std::sync::Arc;
use std::sync::mpsc;
use std::thread;
use clap::{Parser, Arg};

#[derive(Parser)]
#[command(name = "video_compress")]
struct Args {
    #[arg(help = "Входной файл или папка")]
    source: String,
    #[arg(short = 'o', long)]
    output: Option<String>,
    #[arg(short = 'c', long, default_value = "h264")]
    codec: String,
    #[arg(long, default_value = "aac")]
    audio_codec: String,
    #[arg(long)]
    crf: Option<u8>,
    #[arg(long)]
    video_bitrate: Option<String>,
    #[arg(long)]
    audio_bitrate: Option<String>,
    #[arg(long)]
    size: Option<String>,
    #[arg(long)]
    fps: Option<u32>,
    #[arg(long, default_value = "medium")]
    preset: String,
    #[arg(long)]
    two_pass: bool,
    #[arg(long)]
    normalize: bool,
    #[arg(long)]
    subtitles: Option<String>,
    #[arg(long)]
    recursive: bool,
    #[arg(long, default_value_t = 4)]
    threads: usize,
    #[arg(long)]
    overwrite: bool,
    #[arg(long)]
    formats: bool,
}

const SUPPORTED_INPUT: &[&str] = &[".mp4", ".mkv", ".avi", ".mov", ".wmv", ".flv", ".webm", ".m4v", ".3gp", ".ogv"];
const VIDEO_CODECS: &[(&str, &str)] = &[("h264", "libx264"), ("hevc", "libx265"), ("vp9", "libvpx-vp9"), ("av1", "libaom-av1")];
const AUDIO_CODECS: &[&str] = &["aac", "mp3", "opus", "ac3"];
const PRESETS: &[&str] = &["ultrafast", "superfast", "veryfast", "faster", "fast", "medium", "slow", "slower", "veryslow"];

fn check_ffmpeg() {
    if Command::new("ffmpeg").arg("-version").output().is_err() {
        eprintln!("Ошибка: ffmpeg не найден. Установите FFmpeg и добавьте в PATH.");
        std::process::exit(1);
    }
}

fn get_video_codec(codec: &str) -> &str {
    for (k, v) in VIDEO_CODECS {
        if *k == codec {
            return v;
        }
    }
    "libx264"
}

fn build_ffmpeg_cmd(input: &str, output: &str, codec: &str, audio_codec: &str,
                    crf: Option<u8>, video_bitrate: Option<&str>, audio_bitrate: Option<&str>,
                    size: Option<&str>, fps: Option<u32>, preset: &str,
                    normalize: bool, subtitles: Option<&str>, pass: Option<u8>) -> Vec<String> {
    let mut cmd = vec!["ffmpeg".to_string(), "-i".to_string(), input.to_string(), "-y".to_string()];
    cmd.push("-c:v".to_string());
    cmd.push(get_video_codec(codec).to_string());
    cmd.push("-c:a".to_string());
    cmd.push(audio_codec.to_string());
    if let Some(c) = crf {
        cmd.push("-crf".to_string());
        cmd.push(c.to_string());
    } else if let Some(vb) = video_bitrate {
        cmd.push("-b:v".to_string());
        cmd.push(vb.to_string());
    }
    if let Some(ab) = audio_bitrate {
        cmd.push("-b:a".to_string());
        cmd.push(ab.to_string());
    }
    if let Some(sz) = size {
        cmd.push("-vf".to_string());
        cmd.push(format!("scale={}:flags=lanczos", sz));
    }
    if let Some(f) = fps {
        cmd.push("-r".to_string());
        cmd.push(f.to_string());
    }
    if PRESETS.contains(&preset) {
        cmd.push("-preset".to_string());
        cmd.push(preset.to_string());
    }
    if normalize {
        cmd.push("-af".to_string());
        cmd.push("loudnorm=I=-16:LRA=11:TP=-1.5".to_string());
    }
    if let Some(sub) = subtitles {
        cmd.push("-vf".to_string());
        cmd.push(format!("subtitles={}", sub));
    }
    if let Some(p) = pass {
        cmd.push("-pass".to_string());
        cmd.push(p.to_string());
        if p == 1 {
            cmd.push("-f".to_string());
            cmd.push("null".to_string());
            cmd.push("/dev/null".to_string());
        }
    }
    cmd.push(output.to_string());
    cmd
}

fn convert_file(input: &str, output: &str, codec: &str, audio_codec: &str,
                crf: Option<u8>, video_bitrate: Option<&str>, audio_bitrate: Option<&str>,
                size: Option<&str>, fps: Option<u32>, preset: &str,
                two_pass: bool, normalize: bool, subtitles: Option<&str>) -> bool {
    if two_pass && video_bitrate.is_some() {
        let pass1_cmd = build_ffmpeg_cmd(input, "/dev/null", codec, audio_codec, crf,
                                         video_bitrate, audio_bitrate, size, fps, preset,
                                         normalize, subtitles, Some(1));
        if Command::new(&pass1_cmd[0]).args(&pass1_cmd[1..]).status().is_err() {
            return false;
        }
        let pass2_cmd = build_ffmpeg_cmd(input, output, codec, audio_codec, crf,
                                         video_bitrate, audio_bitrate, size, fps, preset,
                                         normalize, subtitles, Some(2));
        Command::new(&pass2_cmd[0]).args(&pass2_cmd[1..]).status().is_ok()
    } else {
        let cmd = build_ffmpeg_cmd(input, output, codec, audio_codec, crf,
                                   video_bitrate, audio_bitrate, size, fps, preset,
                                   normalize, subtitles, None);
        Command::new(&cmd[0]).args(&cmd[1..]).status().is_ok()
    }
}

fn find_video_files(root: &str, recursive: bool) -> Vec<PathBuf> {
    let mut files = Vec::new();
    let path = Path::new(root);
    if path.is_file() {
        if let Some(ext) = path.extension() {
            if SUPPORTED_INPUT.contains(&ext.to_str().unwrap_or("").to_lowercase().as_str()) {
                files.push(path.to_path_buf());
            }
        }
        return files;
    }
    if !path.is_dir() {
        return files;
    }
    if recursive {
        for entry in walkdir::WalkDir::new(path) {
            if let Ok(entry) = entry {
                if entry.file_type().is_file() {
                    if let Some(ext) = entry.path().extension() {
                        let ext_lower = ext.to_str().unwrap_or("").to_lowercase();
                        if SUPPORTED_INPUT.contains(&ext_lower.as_str()) {
                            files.push(entry.path().to_path_buf());
                        }
                    }
                }
            }
        }
    } else {
        if let Ok(entries) = fs::read_dir(path) {
            for entry in entries.flatten() {
                let p = entry.path();
                if p.is_file() {
                    if let Some(ext) = p.extension() {
                        let ext_lower = ext.to_str().unwrap_or("").to_lowercase();
                        if SUPPORTED_INPUT.contains(&ext_lower.as_str()) {
                            files.push(p);
                        }
                    }
                }
            }
        }
    }
    files
}

fn main() {
    let args = Args::parse();

    if args.formats {
        println!("Входные форматы: {}", SUPPORTED_INPUT.join(", "));
        println!("Видеокодеки: {}", VIDEO_CODECS.iter().map(|(k,_)| *k).collect::<Vec<_>>().join(", "));
        println!("Аудиокодеки: {}", AUDIO_CODECS.join(", "));
        println!("Пресеты: {}", PRESETS.join(", "));
        return;
    }

    check_ffmpeg();

    if args.crf.is_some() && args.video_bitrate.is_some() {
        eprintln!("Ошибка: укажите либо --crf, либо --video-bitrate, но не оба.");
        std::process::exit(1);
    }

    let files = find_video_files(&args.source, args.recursive);
    if files.is_empty() {
        println!("Не найдено видеофайлов в {}", args.source);
        std::process::exit(1);
    }

    let out_dir = match args.output {
        Some(ref o) => {
            if Path::new(o).is_dir() {
                o.clone()
            } else {
                Path::new(o).parent().unwrap_or(Path::new(".")).to_str().unwrap().to_string()
            }
        }
        None => "./compressed".to_string(),
    };
    fs::create_dir_all(&out_dir).unwrap();
    let total = files.len();
    println!("Найдено {} видеофайлов.", total);
    let (tx, rx) = mpsc::channel();
    let mut handles = vec![];

    let args = Arc::new(args);

    for chunk in files.chunks((total + args.threads - 1) / args.threads) {
        let chunk = chunk.to_vec();
        let tx = tx.clone();
        let out_dir = out_dir.clone();
        let args = args.clone();
        handles.push(thread::spawn(move || {
            for (i, input_file) in chunk.iter().enumerate() {
                let rel = input_file.strip_prefix(&args.source).unwrap_or(input_file);
                let out_path = Path::new(&out_dir).join(rel.with_extension("mp4"));
                if out_path.exists() && !args.overwrite {
                    tx.send((i+1, format!("{} уже существует, пропуск.", out_path.display()))).unwrap();
                    continue;
                }
                if let Some(parent) = out_path.parent() {
                    fs::create_dir_all(parent).unwrap();
                }
                let ok = convert_file(
                    input_file.to_str().unwrap(),
                    out_path.to_str().unwrap(),
                    &args.codec,
                    &args.audio_codec,
                    args.crf,
                    args.video_bitrate.as_deref(),
                    args.audio_bitrate.as_deref(),
                    args.size.as_deref(),
                    args.fps,
                    &args.preset,
                    args.two_pass,
                    args.normalize,
                    args.subtitles.as_deref(),
                );
                if ok {
                    tx.send((i+1, format!("Конвертация {} -> {}", input_file.display(), out_path.display()))).unwrap();
                } else {
                    tx.send((i+1, format!("Ошибка при конвертации {}", input_file.display()))).unwrap();
                }
            }
        }));
    }
    drop(tx);
    let mut count = 0;
    let mut success = 0;
    for (idx, msg) in rx {
        count += 1;
        println!("[{}/{}] {}", count, total, msg);
        if !msg.contains("ошибка") && !msg.contains("пропуск") {
            success += 1;
        }
    }
    for h in handles {
        h.join().unwrap();
    }
    println!("Готово! Успешно: {}, Всего: {}", success, total);
}
